### DISCUSSION TITLE
[Proposal]: Allow skipping constructor of Positional Records

### DISCUSSION THREAD

**Discussion by SWATOPLUS** [dotnet/csharplang#4178] [Language Ideas]
# Allow skipping constructor of Positional Records

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

When I use some Positional Record I want to be able initialize propertities using object initializer.

```C#
public sealed record SequenceStats(double Average, double Median, double StandardDeviation, double Minimum, double Maximum);

var scores = new [] { 1, 2, 3 };

var static = new SequenceStats // CS7036: There is no argument given that corresponds to the required formal parameter 'Average' of 'SequenceStats.SequenceStats(double, double, double, double, double)
{
    Average = scores.Average(), 
    Median = scores.Median(),
    StandardDeviation = scores.StandardDeviation(),
    Minimum = scores.Min(),
    Maximum = scores.Max(),
};
```

## Motivation
[motivation]: #motivation
This can improve readability of object construction.

## Detailed design
[design]: #detailed-design
1. Compiler should move values from property initializers to contructor or just put in constructor default values.
2. Compiler should give some error if constructor skipped and some parameter is not initialized.

## Drawbacks
[drawbacks]: #drawbacks
None

## Alternatives
[alternatives]: #alternatives
```C#
var static = new SequenceStats(0, 0, 0, 0, 0)
{
    Average = scores.Average(), 
    Median = scores.Median(),
    StandardDeviation = scores.StandardDeviation(),
    Minimum = scores.Min(),
    Maximum = scores.Max(),
};
```

## Unresolved questions
[unresolved]: #unresolved-questions
Why it didn't shipped in C#9.0?

**Comment by HaloFour** [comment]
> Why it didn't shipped in C#9.0?

Because it was intentionally designed to work like this.  It's one of the defining features of a positional record.  Nominal records exist for those cases where you don't want a required constructor.

**Reply by SWATOPLUS** [reply]
It is match very good with: [Proposal] Required Properties #3630 
For Positional Records all properties is required, so we are able to skip constructor.

**Reply by HaloFour** [reply]
That's not the only reason for the required constructor.  Positional records are intentionally positional data constructs.  The order of the properties in the constructor matters and the method of construction is intentionally symmetrical with deconstruction.

You are free to define a secondary constructor which calls the primary constructor if that is what you want.  It does require more code, and it should, because it is an abusive use of positional records.  If you don't want the behavior of positional records I'd suggest not using positional records.

```cs
public sealed record SequenceStats(double Average, double Median, double StandardDeviation, double Minimum, double Maximum) {
    public SequenceStats() : this(default, default, default, default, default) { }
}
```

**Reply by SWATOPLUS** [reply]
I wouldn't like to write additional constructor, I want to define many single line types.  I want to use initializer instead of constructor.

It looks very ugly that you are able to write some arguments to constructor and then override it by initializer. Isn't it?

**Reply by HaloFour** [reply]
> I wouldn't like to write additional constructor, I want to define many single line types. I want to use initializer instead of constructor.

The team has suggested looking for a syntax that makes writing nominal records simpler so that you don't have to define `init` properties manually.  However there are questions as to whether or not such syntax is worthwhile and considering such members would still be semicolon-delimited whether that qualifies as "single line" depends on your stylistic preferences.

Either way, the positional construction of positional records is a very intentional design choice for those types.  Making them non-positional would make no sense.  I'd suggest that effort would be better spent making nominal records "easier" to declare.

**Comment by castholm** [comment]
If the goal is to improve readability, wouldn't using the standard constructor with named arguments work?
```csharp
var stats = new SequenceStats(
    Average: scores.Average(), 
    Median: scores.Median(),
    StandardDeviation: scores.StandardDeviation(),
    Minimum: scores.Min(),
    Maximum: scores.Max());
```
This is just as readable using the object initializer, and with named arguments you can even put the arguments in whichever order you prefer.
